Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,789 @@
+    @usableFromInline
+    /* fileprivate */ internal final class Storage: @unchecked Sendable {
+        /// The lock that protects our state.
+        @usableFromInline
+        /* private */ internal let _lock = Lock()
+        /// The counter used to assign an ID to all our yields.
+        @usableFromInline
+        /* private */ internal let _yieldIDCounter = ManagedAtomic<UInt64>(0)
+        /// The state machine.
+        @usableFromInline
+        /* private */ internal var _stateMachine: StateMachine
+        /// The delegate.
+        @usableFromInline
+        /* private */ internal var _delegate: Delegate?
+
+        @usableFromInline
+        /* fileprivate */ internal init(
+            isWritable: Bool,
+            delegate: Delegate
+        ) {
+            self._stateMachine = .init(isWritable: isWritable)
+            self._delegate = delegate
+        }
+
+        @inlinable
+        /* fileprivate */ internal func writerDeinitialized() {
+            let delegate: Delegate? = self._lock.withLock {
+                let action = self._stateMachine.writerDeinitialized()
+
+                switch action {
+                case .callDidTerminate:
+                    let delegate = self._delegate
+                    self._delegate = nil
+
+                    return delegate
+
+                case .none:
+                    return nil
+                }
+            }
+
+            delegate?.didTerminate(failure: nil)
+        }
+
+        @inlinable
+        /* fileprivate */ internal func setWritability(to writability: Bool) {
+            // We are manually locking here since we need to use both the delegate and the
+            // suspendedYields. Doing this with a withLock becomes very unhandy.
+            self._lock.lock()
+            let action = self._stateMachine.setWritability(to: writability)
+
+            switch action {
+            case .callDidYieldAndResumeContinuations(let suspendedYields):
+                let delegate = self._delegate
+                self._lock.unlock()
+
+                delegate?.didYield(contentsOf: suspendedYields.lazy.map { $0.elements }.flatMap { $0 })
+                suspendedYields.forEach { $0.continuation.resume() }
+
+            case .none:
+                self._lock.unlock()
+                return
+            }
+        }
+
+        @inlinable
+        /* fileprivate */ internal func yield<S: Sequence>(contentsOf sequence: S) async throws where S.Element == Element {
+            // Using relaxed is fine here since we do not need any strict ordering just a
+            // unique ID for every yield.
+            let yieldID = self._yieldIDCounter.loadThenWrappingIncrement(ordering: .relaxed)
+
+            try await withTaskCancellationHandler {
+                // We are manually locking here to hold the lock across the withCheckedContinuation call
+                self._lock.lock()
+
+                let action = self._stateMachine.yield(yieldID: yieldID)
```

Comment:
What's the benefit of having the `yieldIDCounter` as an atomic, if we acquire a lock afterwards anyway?